Update dealing-with-latency.md (#1055)
Unity-Technologies/com.unity.multiplayer.docs: docs/learn/dealing-with-latency.md

<!-- TODO Tentacular, if interract with other server driven elements, need these elements to be predicted as well -->
<!-- TODO add diagram examples (stun grenade for example) and flow of reconciliation -->

> :::info

Reviewer: You're removing this?

Author: Unintentional. Added back.

> It can be hard to handle misprediction corrections elegantly and without taking your players out of their immersion. Clients also can't always stay in sync tick by tick, frame by frame. You can adapt your gameplay and animations to allow your clients to desync for a few seconds to resync again to smooth correction handling. By using controlled desyncs, clients can resolve and reinterpret how they reach a common state independently based on their local context. You can consider this technique as a "smart context-aware interpolation" or an "elegant correction" when predicting. It reuses some of [action anticipation](#action-anticipation)'s technique where you use non-gameplay-important animations to hide discrepancies caused by lag and reinterpret how the client will reach the target authoritative state.

Reviewer: ```suggestion
It can be hard to handle misprediction corrections elegantly and without taking your players out of their immersion. Clients also can't always stay in sync tick by tick, frame by frame. You can adapt your gameplay and animations to allow your clients to desync for a few seconds, as long as they reach a common state and resync again to make corrections seamless. By using controlled desyncs, clients can resolve and reinterpret how they reach a common state independently based on their local context. You can consider this technique as a "smart context-aware interpolation" or an "elegant correction" when predicting. It reuses some of [action anticipation](#action-anticipation)'s technique where you use non-gameplay-important animations to hide discrepancies caused by lag and reinterpret how the client will reach the target authoritative state.
```
I think this flows better from a technical point of view? What do you think?

> For example, you could have an assassination mechanic that you want to smoothly handle misprediction correction. You could have your client tell the server "I killed that target," have the server validate to prevent cheating (like with server rewind), and still give advantage to the attacker.

Reviewer: ```suggestion
For example, you could have an assassination mechanic that you want to seamlessly handle misprediction correction. You could have your client tell the server "I killed that target," have the server validate to prevent cheating (like with server rewind), and still give advantage to the attacker.
```
I think "seamless" fits more in this technique. The goal is not to smooth things out like a normal interpolation, but to actually add smart handling with animations to hide it completely, making it seamless.

> In this example, _Alice the Assassin_ is chasing _Bob the Victim_. _Bob_ enters a room and quickly closes the door behind them at time _t=0_ seconds. Simultaneously at _t=0_, _Alice_ hits the **kill** button. Both actions are client-driven and reach the server at the same time, so the server would tell all clients "Alice killed Bob" at _t=0.1_. _Bob_ receives the "You got killed" message at _t=0.2_. However, _Bob_ closed their door at _t=0_, it would appear that _Alice_ killed _Bob_ through the door.

Reviewer: ```suggestion
In this example, _Alice the Assassin_ is chasing _Bob the Victim_. _Bob_ enters a room and quickly closes the door behind them at time _t=0_ seconds. Simultaneously at _t=0_, _Alice_ hits the **kill** button. Both actions are client-driven and reach the server at the same time, so the server would tell all clients "Alice killed Bob" at _t=0.1_. _Bob_ receives the "You got killed" message at _t=0.2_. However, _Bob_ locally closed their door at _t=0_, it would appear that _Alice_ killed _Bob_ through the door.
```
I'm putting back the word "locally" as it helps figure out whether bob closed the door on their client, on the server, on someone else's client.

One solution is to use server authoritative gameplay with prediction and correction on both clients as described in other sections on this page. The door is corrected by teleporting back open, and the killing animations would play as usual. That correction would remove your players from their immersion and frustrate _Bob_.

> Another solution is to use action anticipation on the door closing and have the door be server driven. You remove the correction and desync problems altogether. However, it still adds lag to each door opening and closing.

Reviewer: In my original text, I mentioned "though if you want your action to be instant with no animation to hide lag." since somewhere else in this doc we talk about action anticipation and how it can help with removing some of the lag perception.
Action anticipation doesn't work with instant actions. If I want to click a button and the door "teleports open" with no swinging animations, lag will still affect my gameplay. Controlled Desyncs allows to have a controlled desync and have that door be client driven, without waiting for the server.

Author: Gotcha. I definitely misread the statement, but I understand now.
